pkg/webhook/resources/virtualmachineimage/validator.go

```diff
@@ -103,6 +114,78 @@ func (v *virtualMachineImageValidator) CheckImageDisplayNameAndURL(newImage *v1b
 	return nil
 }
 
+func (v *virtualMachineImageValidator) checkImageSecurityParameters(newImage *v1beta1.VirtualMachineImage) error {
+	if newImage.Spec.SourceType != v1beta1.VirtualMachineImageSourceTypeClone {
+		return nil
+	}
+
+	if newImage.Spec.SecurityParameters == nil {
+		return werror.NewInvalidError(`sourceParameters is required when image source type is "clone"`, "spec.sourceParameters")
+	}
+
+	if newImage.Spec.SecurityParameters.CryptoOperation == "" {
```

Comment:
I am just curious, do we need to validate the fields on our own if the CRD defines these fields required?

Reply:
It's a very good question. In my original code, I remove all empty checking. But, when I'm writing the unit test, I found the CRD check doesn't work (obviously). So, I added them back.

After testing some order of checking, the admission webhook will check the fields first. Next CRD will check them. So, guarding some request in admission webhook might be a choice?